Lifecycle customization (#592)
DAGWorks-Inc/hamilton: docs/concepts/customizing-execution.rst

            """This function builds the result given the computed values."""
            pass

So we have a few implementations see :doc:`../../reference/result-builders/index` for the list. These can be run independently,
or paired with a graph adapter, see next section for details!

Graph Adapters
##############

Graph Adapters `adapt` the Hamilton DAG, and change how it is executed. They all implement a single interface called
``base.HamiltonGraphAdapter``. They are called internally by Hamilton at the right points in time to make execution
work. The link with the Result Builders, is that GraphAdapters (currently) need to implement a ``build_result()`` function
themselves.

.. code-block:: python

>     class HamiltonGraphAdapter(customization.ResultBuilder):

Reviewer: I think `lifecycle` is a better name than `customization`?

Author: Lifecycle is nebulous -- customization is what we're doing?

Author: Changed to `lifecycle` (we discussed offline) -- more consistent and descriptive enough

        """Any GraphAdapters should implement this interface to adapt the HamiltonGraph for that particular context.

        Note since it inherits ResultMixin -- HamiltonGraphAdapters need a `build_result` function too.
        """
        # four functions not shown

> The default performance is to join all data in a dictionary (unless you bypass `driver.Builder`), in which case it will

Reviewer: ```suggestion
The default behavior is to join all data in a dictionary (unless you bypass `driver.Builder`), in which case it will
```

return a pandas dataframe. Should you find aspects you wish to customize, reach out to the maintainers. The odds are high

> that we already have the right abstraction and have not yet exposed it to the user.

Reviewer: ```suggestion
that we already have the right abstraction and have not yet exposed it via documentation & examples.
```
